src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlUtil.cs

```diff
@@ -2131,4 +2132,48 @@ public static MethodInfo GetPromotedToken
             }
         }
     }
+
+    /// <summary>
+    /// This class implements a FIFO Queue with SemaphoreSlim for ordered execution of parallel tasks.
+    /// Currently used in Managed SNI (SNISslStream) to override SslStream's WriteAsync implementation.
+    /// </summary>
+    internal class ConcurrentQueueSemaphore
+    {
+        private readonly SemaphoreSlim _semaphore;
+        private readonly ConcurrentQueue<TaskCompletionSource<bool>> _queue =
+            new ConcurrentQueue<TaskCompletionSource<bool>>();
+
+        public ConcurrentQueueSemaphore(int initialCount)
+        {
+            _semaphore = new SemaphoreSlim(initialCount);
+        }
+
+        public ConcurrentQueueSemaphore(int initialCount, int maxCount)
+        {
+            _semaphore = new SemaphoreSlim(initialCount, maxCount);
+        }
+
+        public void Wait()
+        {
+            WaitAsync().Wait();
+        }
+
+        public Task WaitAsync()
+        {
+            var tcs = new TaskCompletionSource<bool>();
+            _queue.Enqueue(tcs);
+            _semaphore.WaitAsync().ContinueWith(t =>
+            {
+                if (_queue.TryDequeue(out TaskCompletionSource<bool> popped))
+                    popped.SetResult(true);
+            });
+            return tcs.Task;
+        }
```

Comment:
Removed queue as it's not needed and SemaphoreSlim is concurrent.

Reply:
If you pass in `_queue` as state you avoid the async closure allocation.

Reply:
I'm curious about the correctness of the order here.
Assume a condition that there are 5 items in the queue that are waiting until the semaphore receives a signal; if the 5th `semaphore` receives the signal before others, then it will dequeue the first `tcs` and set its value and will return the 5th `tcs.Task` while the first item is still waiting to receive a signal to unblock.
Any thoughts?

@roji , @Wraith2

Reply:
I'm not sure if the conversation is going forward on the same page. If I'm wrong please skip this comment.

Semaphore doesn't support the order and in this case, we need to guarantee the order besides supporting the concurrency: Please, follow the last line in the remark section in [MS documents SemaphoreSlim](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=net-5.0#remarks).

Please, look at the conversation with Shay [here](https://github.com/dotnet/SqlClient/pull/579#discussion_r493018970).

Reply:
I think the initial design I proposed when calling ContinueWith on WaitAsync was going to preserve the order, as it would continue the read/write operation on the same task from queue without losing handle. With async/await it would look something like this:

```cs
public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
{
    try
    {
        return await await _readAsyncSemaphore.WaitAsync().ContinueWith(t => base.ReadAsync(buffer, offset, count, cancellationToken));
    }
    finally
    {
        _readAsyncSemaphore.Release();
    }
}
```

Let me know what you think.
[Tests are passing]

Reply:
It is correct as far as I can see. the use of queue popping ensures that the waits are released in the same order that they arrived in the queue. The semaphore just prevents concurrency.

I'm not aware of any reason that multiple reads would ever be issued at the same time legitimately though. Sending you can get our of band packets but you wouldn't see that as out of band on the receive side it's just another packet. The only thing that should be causing a read is asking for another packet and asking for a read while one is in progress had to be an error in that case doesn't it? So is there a repro where this queue is ever used?

Reply:
@roji Right, I agree the continue is not necessary.. it does the same thing.

> I'm not aware of any reason that multiple reads would ever be issued at the same time legitimately though. 

@Wraith2 yes, since we acquire only 1 semaphore handle - we're good!

> The only thing that should be causing a read is asking for another packet and asking for a read while one is in progress had to be an error in that case doesn't it? So is there a repro where this queue is ever used?

I have a handy code to study how queue executes tasks, out of driver.. in case you're interested: https://gist.github.com/cheenamalhotra/d6d4140b35536064db14d744d4d700c1

Reply:
Remember to sprinkle in those `ConfigureAwaits()` 😁 

I wasn't questioning the functionality of this class, it looks ok to me. I just think it shouldn't be needed because the situation it's guarding against is an error in the first place. Only the parser should cause reads and it's entire purpose is state maintenance so any case where multiple concurrent reads are issued is an error.